# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [3]:
def add_two(x):
    return x+2
print(add_two(3))

# Lambda Function
lambda x: x+2
(lambda x: x+2)(3)

5


5

#### Saving to a Variable

In [4]:
f = lambda x,y : x+y+2
f(3,5)

10

#### Multiple Inputs

In [6]:
#multiple inputs  not stored as a varible
(lambda x,y,z:x*y*z)(3,5,8)

#stored as a varible
x = lambda x,y,z:x*y*z
x(3,5,8)

120

#### Passing a Lambda into a Function

In [7]:
def do_math(f, num):
    return f(num)

multiply(lambda x: x*x,4)


16

#### Returning a Lambda from a Function

In [3]:
def my_func():
    return "jello"

print(my_func())
print(type(my_func()))

jello
<class 'str'>


In [4]:
def return_func():
    def squared(num):
        return num*num
    def divide(num):
        return num/num
    return squared

print(return_func()(3))

9


In [5]:
def return_lamb(b,c):
    return lambda x,y: (b+c)-(x+y)

return_lamb(3,2)(4,3)

-2

#### If Statements within Lambdas

In [13]:
# Lambda x : True if (condition) else False --This is called a Ternary Operator
f = lambda num : num * 2 if num > 10 else num + 2
print(f(10))
print(f(12))

12
24


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [6]:
f = lambda x: x*x*x
f(4)


64

In [16]:
#get good at reading Documentation because this is your lifelinealso
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In [7]:
#Using Lambdas with sort to sort a list of dictionary by a key from highest to smallest
x = [2,56745,56,4,2516,36547,56,3241234,346]
print(sorted(x, reverse = True))
print(x)

[3241234, 56745, 36547, 2516, 346, 56, 56, 4, 2]
[2, 56745, 56, 4, 2516, 36547, 56, 3241234, 346]


In [20]:
# Using lambda with sort to sort a list of dictionaries by a key from highest to lowest
k = {
    "name":"Kevin",
    "age":36
}
j = {
    "name":"John",
    "age":41
}
d = {
    "name":"Derek",
    "age":32
}
b = {
    "name":"Brian",
    "age":29
}

my_list = [k,j,d,b]
print(my_list)

print(
    sorted(my_list, key=lambda person_dict: person_dict["age"])
)

print(
    sorted(my_list, key=lambda person_dict: person_dict["age"], reverse = True)
)


[{'name': 'Kevin', 'age': 36}, {'name': 'John', 'age': 41}, {'name': 'Derek', 'age': 32}, {'name': 'Brian', 'age': 29}]
[{'name': 'Brian', 'age': 29}, {'name': 'Derek', 'age': 32}, {'name': 'Kevin', 'age': 36}, {'name': 'John', 'age': 41}]
[{'name': 'John', 'age': 41}, {'name': 'Kevin', 'age': 36}, {'name': 'Derek', 'age': 32}, {'name': 'Brian', 'age': 29}]


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [8]:
# map(func, iterable)
# help(map)
#typically you will use predefined func buy you can use lambdas as well
help(map)

Help on class map in module builtins:

class map(object)
 |  map(func, *iterables) --> map object
 |  
 |  Make an iterator that computes the function using arguments from
 |  each of the iterables.  Stops when the shortest iterable is exhausted.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



[(16, 16), (10, 11), (3, 20), (4, 9), (6, 15)]

In [ ]:
def squared(num, num2):
    if num < 10 and num2 < 10:
        return num**2, num2**2
    else:
        return num, num2
numbers = [4,10,3,2,6]
more_nums = [4,11,20,3,15]

squared_nums_map = map(squared, numbers, more_nums)
list(squared_nums_map)


#### Using Lambda's with Map

In [9]:
#map(Lambda x+2,list)
# using Lambda in map happens in line (or in one line)
# list(map(lambda x: x*2,nums))
numbers = [4,10,3,2,6]

squared_nums = list(map(lambda x: x**2 if x<10 else x,numbers)) #when passing a lambda function it's good practice to put it in parenthesis
print(squared_nums)

[16, 10, 9, 4, 36]


In [10]:
numbers = [4,10,3,2,6]
more_nums = [4,1,20,3,15]
squared_num2 = list(map((lambda x,y: (x**2, y**2) if x<10 and y<10 else (x,y)), numbers, more_nums))
print(squared_num2)

[(16, 16), (10, 1), (3, 20), (4, 9), (6, 15)]


In [12]:
# Using Lambdas to sort a list of dictionaries by a key from highest to lowest

k = {
    "name":"Kyle",
    "age":12
}

t = {
    "name":"Kevin",
    "age": 27
}

p = {
    "name":"Patrick",
    "age":27
}

people = [k,t,p]
print(people)

my_lambda = lambda person_dict:person_dict['age']
print(sorted(people, key=my_lambda))
print(sorted(people, key=my_lambda, reverse=True))

[{'name': 'Kyle', 'age': 12}, {'name': 'Kevin', 'age': 27}, {'name': 'Patrick', 'age': 27}]
[{'name': 'Kyle', 'age': 12}, {'name': 'Kevin', 'age': 27}, {'name': 'Patrick', 'age': 27}]
[{'name': 'Kevin', 'age': 27}, {'name': 'Patrick', 'age': 27}, {'name': 'Kyle', 'age': 12}]


#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [17]:
nums = [4,2,6,8]

double = list(map((lambda x: (x*2)-1), nums))
print(double)

[7, 3, 11, 15]


## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [45]:
help(filter)

Help on class filter in module builtins:

class filter(object)
 |  filter(function or None, iterable) --> filter object
 |  
 |  Return an iterator yielding those items of iterable for which function(item)
 |  is true. If function is None, return the items that are true.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



In [18]:
names = ["Dak", "Ramika", "Armani", "Ku", "Gareth", "Kyle", "Kevin"]

# we only want names that start with K

def k_names(name):
    if name[0].lower()=="k":
        return True
    else:
        return False

new_names_list = list(filter(k_names,names))
new_names_list

['Ku', 'Kyle', 'Kevin']

#### Using Lambda's with Filter()

In [47]:
new_names_lamb = list(filter(lambda name: True if name[0].lower()=='k' else False, names))
new_names_lamb

['Andy', 'Angelica']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [22]:
from statistics import mean
# help(mean)
nums = [2,7,4.2,1.6,9,4.4,4.9]

mean_nums = mean(nums)
new_nums = list(filter(lambda x: True if x <= mean_nums else False, nums))
print(new_nums)


[2, 4.2, 1.6, 4.4]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [24]:
from functools import reduce
help(reduce)

Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, sequence[, initial]) -> value
    
    Apply a function of two arguments cumulatively to the items of a sequence,
    from left to right, so as to reduce the sequence to a single value.
    For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
    ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
    of the sequence in the calculation, and serves as a default when the
    sequence is empty.



In [55]:
from functools import reduce
# reduce(function, iterable)
# help(reduce)

list_1 = [2,4,6,8,10]

def subtract_nums(num1, num2):
    return num1 - num2
result = reduce(subtract_nums,list_1)
result


-26

#### Using Lambda's with Reduce()

In [56]:
l_3 = [1,10,2,30]
                        #x is aggregator and y is the list value
result_lamb = reduce(lambda x,y :x-y,l_3)
print(result_lamb)

-41


#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [28]:
nums  = [2,5,8,34,50]

result_lamb = reduce(lambda x,y :x*y,nums)
print(result_lamb)


136000


## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [31]:
def recursive(i):
    print("hi")
    if i == 4: #Base Case
        return
    else:
        i += 1
        recursive(i)
    
recursive(0)

hi
hi
hi
hi
hi


#### Implementing a Base Case

In [33]:
def add_nums(num):
    if num <=1:
        print("add_nums of 1 = 1")
        return num
    else:
        print(f"add_nums({num}) = {num} + add_nums({num-1})")
        x = num -1
        return num + add_nums(x)

add_nums(5)


add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums of 1 = 1


15

#### Writing a Factorial Function

In [34]:
#5! = 5 * 4 * 3 * 2 * 1

def factorial(num):
    #set base case:
    if num <= 1:
        print('factorial(1) = 1')
        return 1
    else:
        print(f'factorial({num}) = {num} * factorial({num-1})')
        return num * factorial(num - 1)
factorial(50)

factorial(50) = 50 * factorial(49)
factorial(49) = 49 * factorial(48)
factorial(48) = 48 * factorial(47)
factorial(47) = 47 * factorial(46)
factorial(46) = 46 * factorial(45)
factorial(45) = 45 * factorial(44)
factorial(44) = 44 * factorial(43)
factorial(43) = 43 * factorial(42)
factorial(42) = 42 * factorial(41)
factorial(41) = 41 * factorial(40)
factorial(40) = 40 * factorial(39)
factorial(39) = 39 * factorial(38)
factorial(38) = 38 * factorial(37)
factorial(37) = 37 * factorial(36)
factorial(36) = 36 * factorial(35)
factorial(35) = 35 * factorial(34)
factorial(34) = 34 * factorial(33)
factorial(33) = 33 * factorial(32)
factorial(32) = 32 * factorial(31)
factorial(31) = 31 * factorial(30)
factorial(30) = 30 * factorial(29)
factorial(29) = 29 * factorial(28)
factorial(28) = 28 * factorial(27)
factorial(27) = 27 * factorial(26)
factorial(26) = 26 * factorial(25)
factorial(25) = 25 * factorial(24)
factorial(24) = 24 * factorial(23)
factorial(23) = 23 * factorial(22)
factorial(22) = 22 *

30414093201713378043612608166064768844377641568960512000000000000

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [38]:
def sub_nums(num):
    if num <= 1:
        print('sub_nums of 1 = 1')
        return num
    else:
        print(f"sub_nums({num}) = {num} - sub_nums({num-1})")
        x = num - 1
        return num - sub_nums(x)
    
sub_nums(10)

sub_nums(10) = 10 - sub_nums(9)
sub_nums(9) = 9 - sub_nums(8)
sub_nums(8) = 8 - sub_nums(7)
sub_nums(7) = 7 - sub_nums(6)
sub_nums(6) = 6 - sub_nums(5)
sub_nums(5) = 5 - sub_nums(4)
sub_nums(4) = 4 - sub_nums(3)
sub_nums(3) = 3 - sub_nums(2)
sub_nums(2) = 2 - sub_nums(1)
sub_nums of 1 = 1


5

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

In [69]:
range(0,10)
for i in range(0,10):
    print(i)

0
1
2
3
4
5
6
7
8
9


#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [71]:
def my_range(stop, start, step = 1):
    while start < stop:
        yield start
        start += step
for i in my_range(20,2):
    print(i)

2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19


#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [74]:
def range_squared(stop, start, step =1):
    while start > stop:
        yield start**2
        start -= step
        
for num in range_squared(0,10,2):
    print(num)

100
64
36
16
4


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [64]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]
new_places_lamb = list(filter(lambda place: True if place.strip() != "" else False, places))
print(new_places_lamb)



['Argentina', 'San Diego', 'Boston', 'New York']


### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [66]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

author.sort(key=lambda name: name.split(" ")[-1].lower())
print(author)
    


['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [110]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]

c_to_f = list(map(lambda place: (place[0], (9/5)*place[1] + 32), places))
print(c_to_f)

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [114]:
def fib_is_evil(num):
    if num <= 1:
        return 1
    else:
        return(fib_is_evil(num-1) + fib_is_evil(num-2))
    
def pretty_fib(num1):
    for i in range(num1 + 1):
        
        print(f"Iteration {i} : {fib_is_evil(i)}")

pretty_fib(10)

Iteration 0 : 1
Iteration 1 : 1
Iteration 2 : 2
Iteration 3 : 3
Iteration 4 : 5
Iteration 5 : 8
Iteration 6 : 13
Iteration 7 : 21
Iteration 8 : 34
Iteration 9 : 55
Iteration 10 : 89
